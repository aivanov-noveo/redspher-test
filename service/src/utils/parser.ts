import { parse } from '@babel/parser';
import type { BinaryExpression, ExpressionStatement } from '@babel/types';

/**
 * A parsing result.
 * @field expression a parsed binary expression,
 * is present only if there is no parsing error, optional
 * @field error an occurred error message (if any),
 * is present only if there is a parsing error, optional
 */
export interface ParseResult {
  expression?: BinaryExpression;
  error?: string;
}

const EXPRESSION_STATEMENT = 'ExpressionStatement';
const BINARY_EXPRESSION = 'BinaryExpression';

/**
 * Parses an arithmetic expression string into AST.
 * The expression should contain at least one binary operation.
 * @example
 * 1.5 + 2*3 - 5/2 + +2 - -1
 * @param expression An arithmetic expression string.
 * @returns a parsing result with either AST contains numeric literal, unary operation
 * and binary operation nodes or a parsing error.
 * @example
 * ParseResult:
 * {
 *   // a parsed binary expression root node from AST, optional
 *   expression: BinaryExpression,
 *   // a parsing error (if any), optional
 *   error: string
 * }
 */
export const parseAST = (expression: string): ParseResult => {
  try {
    const ast = parse(expression);

    if (ast.errors.length > 0) {
      return {
        error: 'Unknown error',
      };
    }

    const statement = ast.program.body[0];

    if (statement.type !== EXPRESSION_STATEMENT) {
      return {
        error: 'Incorrect expression',
      };
    }

    const expressionStatement = statement as ExpressionStatement;

    if (expressionStatement.expression.type !== BINARY_EXPRESSION) {
      return {
        error: 'Incorrect expression',
      };
    }

    return {
      expression: expressionStatement.expression,
    };
  } catch (error: unknown) {
    console.error(error);

    return {
      error: 'Syntax error',
    };
  }
};
