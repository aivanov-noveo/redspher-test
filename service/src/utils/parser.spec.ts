import { parseAST } from './parser';
import * as parserModule from '@babel/parser';
import { Program } from '@babel/types';

describe('parseAST', () => {
  const parseSpy = jest.spyOn(parserModule, 'parse');
  const consoleErrorSpy = jest.spyOn(console, 'error');
  consoleErrorSpy.mockImplementation();

  it('passes an expression to the parse function', () => {
    expect.assertions(1);

    parseAST('1 + 2');

    expect(parseSpy).toHaveBeenCalledWith('1 + 2');
  });

  it('returns a syntax error if a parser throws SyntaxError', () => {
    expect.assertions(1);

    const result = parseAST('1 + 2..2');

    expect(result.error).toBe('Syntax error');
  });

  it('logs an error to console if a parser throws an error', () => {
    expect.assertions(1);

    parseSpy.mockImplementationOnce((): never => {
      throw new Error('Unknown error');
    });

    parseAST('1 + 2');

    expect(consoleErrorSpy).toHaveBeenCalledWith(new Error('Unknown error'));
  });

  it('returns an unknown error if a parser returns errors', () => {
    expect.assertions(1);

    parseSpy.mockReturnValueOnce({
      errors: [
        {
          code: 'some error code',
          reasonCode: 'reason code',
        },
      ],
      type: 'File',
      program: undefined as unknown as Program,
    });
    const result = parseAST('1 + 2..2');

    expect(result.error).toBe('Unknown error');
  });

  it('returns an incorrect expression error if a parser returns not an expression statement', () => {
    expect.assertions(1);

    parseSpy.mockReturnValueOnce({
      errors: [],
      type: 'File',
      program: {
        body: [
          {
            type: 'BlockStatement',
          },
        ],
      } as unknown as Program,
    });
    const result = parseAST('1 + 2');

    expect(result.error).toBe('Incorrect expression');
  });

  it('returns an incorrect expression error if a parser returns not a binary expression', () => {
    expect.assertions(1);

    parseSpy.mockReturnValueOnce({
      errors: [],
      type: 'File',
      program: {
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'ArrayExpression',
            },
          },
        ],
      } as unknown as Program,
    });
    const result = parseAST('1 + 2');

    expect(result.error).toBe('Incorrect expression');
  });

  it('returns a parsed AST if a correct expression string is passed', () => {
    expect.assertions(1);

    const result = parseAST('1 + 2');

    expect(result.expression).toMatchSnapshot('parsed AST');
  });
});
