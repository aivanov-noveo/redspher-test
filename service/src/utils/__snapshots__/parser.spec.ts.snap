// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAST returns a parsed AST if a correct expression string is passed: parsed AST 1`] = `
Node {
  "end": 5,
  "left": Node {
    "end": 1,
    "extra": Object {
      "raw": "1",
      "rawValue": 1,
    },
    "loc": SourceLocation {
      "end": Position {
        "column": 1,
        "index": 1,
        "line": 1,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "index": 0,
        "line": 1,
      },
    },
    "start": 0,
    "type": "NumericLiteral",
    "value": 1,
  },
  "loc": SourceLocation {
    "end": Position {
      "column": 5,
      "index": 5,
      "line": 1,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 0,
      "index": 0,
      "line": 1,
    },
  },
  "operator": "+",
  "right": Node {
    "end": 5,
    "extra": Object {
      "raw": "2",
      "rawValue": 2,
    },
    "loc": SourceLocation {
      "end": Position {
        "column": 5,
        "index": 5,
        "line": 1,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 4,
        "index": 4,
        "line": 1,
      },
    },
    "start": 4,
    "type": "NumericLiteral",
    "value": 2,
  },
  "start": 0,
  "type": "BinaryExpression",
}
`;
